lib/puppet/settings/symlink_or_directory_setting.rb

```diff
@@ -0,0 +1,14 @@
+class Puppet::Settings::SymlinkOrDirectorySetting < Puppet::Settings::DirectorySetting
+
+  def initialize(args)
+    super
+  end
+
+  def type
+    if Puppet::FileSystem.symlink?(self.value)
+      :symlink
+    else
+      :directory
+    end
```

Comment:
Hi @mwaggett looks good. I had a few questions/comments. Could you add a new unit test in `spec/unit/settings/symlink_or_directory_spec.rb` and verify what happens when trying to manage the setting when the value refers to a non-existent directory, a directory, a symlink to a directory, and a danging symlink? For the last one, should the `type` method return `:symlink` if the symlink is dangling?

Reply:
Will do!

I was still trying to figure out our desired behavior - is the presence of a symlink enough or do we want to read the link and verify it's not dangling? presumably the latter..

Here's some pseudo-code with my thoughts:

Previously: 
```
if codedir is a directory: great
else: attempt to create directory
```

Now:
```
if codedir is a directory: great
elseif codedir is a symlink:
  if symlink resolves: great (does it need to resolve to a dir or is a file okay?)
  else: ??? attempt to create the thing it points to? as a dir?
else: attempt to create directory
```

Reply:
^ lmk if y'all have thoughts @Magisus @justinstoller

Reply:
I think we want what would normally be an `ensure => present` in Puppet, but instead of a creating a file if non-existent create a directory. I don't think we should try to resolve the symlink. What the target of that symlink is isn't a meaningful thing in our current settings. I think if it's a symlink it's on the user to own that correctness.

Would it be possible to just do an `ensure => present`? It's not exactly what we want, but honestly I don't think what we want is possible in Puppet w/o multiple resources.

Reply:
I think you could override the `FileSetting#to_resource` method to return an instance of `Puppet::Resource` with whatever `resource[:ensure] = 'present'`  makes sense.